File: src/resilience/testing.rs
//! Testing tools for resilience patterns
//!
//! This module provides tools for testing resilience patterns,
//! including circuit breakers, bulkheads, retries, and fallbacks.

use anyhow::Result;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Arc;
use std::time::{Duration, SystemTime, UNIX_EPOCH};
use tokio::sync::{Mutex, RwLock};
use tracing::{debug, error, info, warn};

use super::circuit_breaker::{CircuitBreaker, CircuitBreakerConfig, CircuitBreakerState};
use super::bulkhead::Bulkhead;
use super::retry::RetryPolicy;
use super::fallback::FallbackPolicy;

/// Chaos testing configuration
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ChaosTestingConfig {
    /// Failure rate (0.0 - 1.0)
    pub failure_rate: f64,
    /// Latency injection in milliseconds
    pub latency_ms: u64,
    /// Whether to inject latency
    pub inject_latency: bool,
    /// Whether to inject failures
    pub inject_failures: bool,
    /// Whether to inject timeouts
    pub inject_timeouts: bool,
    /// Timeout rate (0.0 - 1.0)
    pub timeout_rate: f64,
    /// Whether to inject connection errors
    pub inject_connection_errors: bool,
    /// Connection error rate (0.0 - 1.0)
    pub connection_error_rate: f64,
}

impl Default for ChaosTestingConfig {
    fn default() -> Self {
        Self {
            failure_rate: 0.2,
            latency_ms: 100,
            inject_latency: false,
            inject_failures: false,
            inject_timeouts: false,
            timeout_rate: 0.1,
            inject_connection_errors: false,
            connection_error_rate: 0.1,
        }
    }
}

/// Circuit breaker test result
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct CircuitBreakerTestResult {
    /// Service name
    pub service_name: String,
    /// Test duration in milliseconds
    pub duration_ms: u64,
    /// Number of requests
    pub request_count: u64,
    /// Number of successful requests
    pub success_count: u64,
    /// Number of failed requests
    pub failure_count: u64,
    /// Number of rejected requests
    pub rejected_count: u64,
    /// Number of state transitions
    pub state_transitions: u64,
    /// Final state
    pub final_state: CircuitBreakerState,
    /// Time spent in each state (milliseconds)
    pub time_in_state: HashMap<CircuitBreakerState, u64>,
    /// Average response time (milliseconds)
    pub avg_response_time_ms: f64,
}

/// Circuit breaker tester
pub struct CircuitBreakerTester {
    /// Circuit breaker
    circuit_breaker: CircuitBreaker,
    /// Chaos testing configuration
    config: Arc<Mutex<ChaosTestingConfig>>,
    /// Random number generator
    rng: Arc<Mutex<rand::rngs::ThreadRng>>,
}

impl CircuitBreakerTester {
    /// Create a new circuit breaker tester
    pub fn new(circuit_breaker: CircuitBreaker, config: ChaosTestingConfig) -> Self {
        Self {
            circuit_breaker,
            config: Arc::new(Mutex::new(config)),
            rng: Arc::new(Mutex::new(rand::thread_rng())),
        }
    }
    
    /// Update chaos testing configuration
    pub async fn update_config(&self, config: ChaosTestingConfig) {
        let mut current_config = self.config.lock().await;
        *current_config = config;
    }
    
    /// Run a test with a specified number of requests
    pub async fn run_test(&self, request_count: u64) -> Result<CircuitBreakerTestResult> {
        let start_time = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_millis() as u64;
        
        let mut success_count = 0;
        let mut failure_count = 0;
        let mut rejected_count = 0;
        let mut total_response_time = 0;
        
        for _ in 0..request_count {
            let request_start_time = SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap_or_default()
                .as_millis() as u64;
            
            // Check if request is allowed
            if !self.circuit_breaker.allow_request().await {
                rejected_count += 1;
                self.circuit_breaker.record_rejected().await?;
                continue;
            }
            
            // Simulate request
            let result = self.simulate_request().await;
            
            let request_end_time = SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap_or_default()
                .as_millis() as u64;
            
            let response_time = request_end_time - request_start_time;
            total_response_time += response_time;
            
            match result {
                Ok(_) => {
                    success_count += 1;
                    self.circuit_breaker.record_success().await?;
                }
                Err(_) => {
                    failure_count += 1;
                    self.circuit_breaker.record_failure().await?;
                }
            }
        }
        
        let end_time = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_millis() as u64;
        
        let duration_ms = end_time - start_time;
        
        let metrics = self.circuit_breaker.get_metrics().await;
        
        Ok(CircuitBreakerTestResult {
            service_name: "test-service".to_string(),
            duration_ms,
            request_count,
            success_count,
            failure_count,
            rejected_count,
            state_transitions: metrics.state_transition_count,
            final_state: self.circuit_breaker.get_state().await,
            time_in_state: metrics.time_in_state,
            avg_response_time_ms: if success_count + failure_count > 0 {
                total_response_time as f64 / (success_count + failure_count) as f64
            } else {
                0.0
            },
        })
    }
    
    /// Simulate a request with chaos injection
    async fn simulate_request(&self) -> Result<(), anyhow::Error> {
        let config = self.config.lock().await.clone();
        let mut rng = self.rng.lock().await;
        
        // Inject latency if enabled
        if config.inject_latency {
            let latency = config.latency_ms;
            tokio::time::sleep(Duration::from_millis(latency)).await;
        }
        
        // Inject connection errors if enabled
        if config.inject_connection_errors && rand::Rng::gen_bool(&mut *rng, config.connection_error_rate) {
            return Err(anyhow::anyhow!("Simulated connection error"));
        }
        
        // Inject timeouts if enabled
        if config.inject_timeouts && rand::Rng::gen_bool(&mut *rng, config.timeout_rate) {
            return Err(anyhow::anyhow!("Simulated timeout"));
        }
        
        // Inject failures if enabled
        if config.inject_failures && rand::Rng::gen_bool(&mut *rng, config.failure_rate) {
            return Err(anyhow::anyhow!("Simulated failure"));
        }
        
        Ok(())
    }
}

/// Resilience test suite
pub struct ResilienceTestSuite {
    /// Circuit breaker tester
    circuit_breaker_tester: Option<CircuitBreakerTester>,
    /// Chaos testing configuration
    config: Arc<Mutex<ChaosTestingConfig>>,
}

impl ResilienceTestSuite {
    /// Create a new resilience test suite
    pub fn new() -> Self {
        Self {
            circuit_breaker_tester: None,
            config: Arc::new(Mutex::new(ChaosTestingConfig::default())),
        }
    }
    
    /// Set circuit breaker tester
    pub fn with_circuit_breaker(mut self, circuit_breaker: CircuitBreaker) -> Self {
        let config = ChaosTestingConfig::default();
        self.circuit_breaker_tester = Some(CircuitBreakerTester::new(circuit_breaker, config));
        self
    }
    
    /// Update chaos testing configuration
    pub async fn update_config(&self, config: ChaosTestingConfig) {
        let mut current_config = self.config.lock().await;
        *current_config = config.clone();
        
        if let Some(tester) = &self.circuit_breaker_tester {
            tester.update_config(config).await;
        }
    }
    
    /// Run a circuit breaker test
    pub async fn run_circuit_breaker_test(&self, request_count: u64) -> Result<CircuitBreakerTestResult> {
        if let Some(tester) = &self.circuit_breaker_tester {
            tester.run_test(request_count).await
        } else {
            Err(anyhow::anyhow!("Circuit breaker tester not configured"))
        }
    }
    
    /// Run a test scenario that verifies circuit breaker behavior
    pub async fn verify_circuit_breaker_behavior(&self) -> Result<bool> {
        if let Some(tester) = &self.circuit_breaker_tester {
            // Configure chaos to inject failures
            tester.update_config(ChaosTestingConfig {
                failure_rate: 1.0, // 100% failure rate
                inject_failures: true,
                ..ChaosTestingConfig::default()
            }).await;
            
            // Run test with enough requests to trigger circuit breaker
            let result = tester.run_test(10).await?;
            
            // Verify circuit breaker opened
            if result.final_state != CircuitBreakerState::Open {
                return Ok(false);
            }
            
            // Configure chaos to stop injecting failures
            tester.update_config(ChaosTestingConfig {
                failure_rate: 0.0,
                inject_failures: false,
                ..ChaosTestingConfig::default()
            }).await;
            
            // Wait for circuit breaker to transition to half-open
            tokio::time::sleep(Duration::from_millis(31000)).await;
            
            // Run test with enough requests to close circuit breaker
            let result = tester.run_test(10).await?;
            
            // Verify circuit breaker closed
            Ok(result.final_state == CircuitBreakerState::Closed)
        } else {
            Err(anyhow::anyhow!("Circuit breaker tester not configured"))
        }
    }
}

/// Resilience test report
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ResilienceTestReport {
    /// Test name
    pub test_name: String,
    /// Test duration in milliseconds
    pub duration_ms: u64,
    /// Test result (success/failure)
    pub success: bool,
    /// Test details
    pub details: HashMap<String, String>,
    /// Circuit breaker test results
    pub circuit_breaker_results: Option<CircuitBreakerTestResult>,
}

/// Resilience test runner
pub struct ResilienceTestRunner {
    /// Test suites
    test_suites: HashMap<String, ResilienceTestSuite>,
    /// Test reports
    test_reports: Arc<RwLock<Vec<ResilienceTestReport>>>,
}

impl ResilienceTestRunner {
    /// Create a new resilience test runner
    pub fn new() -> Self {
        Self {
            test_suites: HashMap::new(),
            test_reports: Arc::new(RwLock::new(Vec::new())),
        }
    }
    
    /// Add a test suite
    pub fn add_test_suite(&mut self, name: &str, suite: ResilienceTestSuite) {
        self.test_suites.insert(name.to_string(), suite);
    }
    
    /// Run all tests
    pub async fn run_all_tests(&self) -> Result<Vec<ResilienceTestReport>> {
        let mut reports = Vec::new();
        
        for (name, suite) in &self.test_suites {
            let start_time = SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap_or_default()
                .as_millis() as u64;
            
            let result = suite.verify_circuit_breaker_behavior().await;
            
            let end_time = SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap_or_default()
                .as_millis() as u64;
            
            let duration_ms = end_time - start_time;
            
            let success = result.is_ok() && result.unwrap();
            
            let mut details = HashMap::new();
            if let Err(e) = &result {
                details.insert("error".to_string(), e.to_string());
            }
            
            let report = ResilienceTestReport {
                test_name: name.clone(),
                duration_ms,
                success,
                details,
                circuit_breaker_results: None,
            };
            
            reports.push(report.clone());
            
            // Store report
            self.test_reports.write().await.push(report);
        }
        
        Ok(reports)
    }
    
    /// Get all test reports
    pub async fn get_test_reports(&self) -> Vec<ResilienceTestReport> {
        self.test_reports.read().await.clone()
    }
    
    /// Clear test reports
    pub async fn clear_test_reports(&self) {
        self.test_reports.write().await.clear();
    }
}

/// Create a circuit breaker for testing
pub fn create_test_circuit_breaker(service_name: &str) -> CircuitBreaker {
    let config = CircuitBreakerConfig {
        consecutive_errors_threshold: 5,
        open_to_half_open_timeout_ms: 30000,
        half_open_success_threshold: 3,
        ..CircuitBreakerConfig::default()
    };
    
    CircuitBreaker::new(service_name.to_string(), config)
}

/// Create a resilience test suite for a service
pub fn create_test_suite_for_service(service_name: &str) -> ResilienceTestSuite {
    let circuit_breaker = create_test_circuit_breaker(service_name);
    
    ResilienceTestSuite::new()
        .with_circuit_breaker(circuit_breaker)
}

File: internal/resilience/testing.go
// Package resilience provides resilience patterns for the Hivemind system.
package resilience

import (
	"context"
	"errors"
	"math/rand"
	"sync"
	"time"

	"go.uber.org/zap"
)

// ChaosTestingConfig defines the configuration for chaos testing
type ChaosTestingConfig struct {
	// FailureRate is the failure rate (0.0 - 1.0)
	FailureRate float64
	// LatencyMs is the latency injection in milliseconds
	LatencyMs int64
	// InjectLatency indicates whether to inject latency
	InjectLatency bool
	// InjectFailures indicates whether to inject failures
	InjectFailures bool
	// InjectTimeouts indicates whether to inject timeouts
	InjectTimeouts bool
	// TimeoutRate is the timeout rate (0.0 - 1.0)
	TimeoutRate float64
	// InjectConnectionErrors indicates whether to inject connection errors
	InjectConnectionErrors bool
	// ConnectionErrorRate is the connection error rate (0.0 - 1.0)
	ConnectionErrorRate float64
}

// DefaultChaosTestingConfig returns the default chaos testing configuration
func DefaultChaosTestingConfig() *ChaosTestingConfig {
	return &ChaosTestingConfig{
		FailureRate:            0.2,
		LatencyMs:              100,
		InjectLatency:          false,
		InjectFailures:         false,
		InjectTimeouts:         false,
		TimeoutRate:            0.1,
		InjectConnectionErrors: false,
		ConnectionErrorRate:    0.1,
	}
}

// CircuitBreakerTestResult contains the result of a circuit breaker test
type CircuitBreakerTestResult struct {
	// ServiceName is the service name
	ServiceName string
	// DurationMs is the test duration in milliseconds
	DurationMs int64
	// RequestCount is the number of requests
	RequestCount int64
	// SuccessCount is the number of successful requests
	SuccessCount int64
	// FailureCount is the number of failed requests
	FailureCount int64
	// RejectedCount is the number of rejected requests
	RejectedCount int64
	// StateTransitions is the number of state transitions
	StateTransitions uint64
	// FinalState is the final state
	FinalState CircuitBreakerState
	// TimeInState is the time spent in each state (milliseconds)
	TimeInState map[CircuitBreakerState]int64
	// AvgResponseTimeMs is the average response time (milliseconds)
	AvgResponseTimeMs float64
}

// CircuitBreakerTester is a tester for circuit breakers
type CircuitBreakerTester struct {
	// CircuitBreaker is the circuit breaker to test
	circuitBreaker *CircuitBreaker
	// Config is the chaos testing configuration
	config *ChaosTestingConfig
	// Random number generator
	rand *rand.Rand
	// Mutex for protecting config updates
	mu sync.Mutex
	// Logger
	logger *zap.Logger
}

// NewCircuitBreakerTester creates a new circuit breaker tester
func NewCircuitBreakerTester(circuitBreaker *CircuitBreaker, config ChaosTestingConfig) *CircuitBreakerTester {
	logger, _ := zap.NewProduction()

	return &CircuitBreakerTester{
		circuitBreaker: circuitBreaker,
		config:         &config,
		rand:           rand.New(rand.NewSource(time.Now().UnixNano())),
		logger:         logger,
	}
}

// UpdateConfig updates the chaos testing configuration
func (t *CircuitBreakerTester) UpdateConfig(config ChaosTestingConfig) {
	t.mu.Lock()
	defer t.mu.Unlock()

	t.config = &config
}

// RunTest runs a test with a specified number of requests
func (t *CircuitBreakerTester) RunTest(ctx context.Context, requestCount int64) (*CircuitBreakerTestResult, error) {
	startTime := time.Now()

	var successCount int64
	var failureCount int64
	var rejectedCount int64
	var totalResponseTime int64

	for i := int64(0); i < requestCount; i++ {
		requestStartTime := time.Now()

		// Check if request is allowed
		if !t.circuitBreaker.AllowRequest() {
			rejectedCount++
			t.circuitBreaker.RecordRejected()
			continue
		}

		// Simulate request
		err := t.simulateRequest(ctx)

		requestEndTime := time.Now()

		responseTime := requestEndTime.Sub(requestStartTime).Milliseconds()
		totalResponseTime += responseTime

		if err == nil {
			successCount++
			t.circuitBreaker.RecordSuccess()
		} else {
			failureCount++
			t.circuitBreaker.RecordFailure()
		}
	}

	endTime := time.Now()

	durationMs := endTime.Sub(startTime).Milliseconds()

	metrics := t.circuitBreaker.GetMetrics()

	return &CircuitBreakerTestResult{
		ServiceName:      "test-service",
		DurationMs:       durationMs,
		RequestCount:     requestCount,
		SuccessCount:     successCount,
		FailureCount:     failureCount,
		RejectedCount:    rejectedCount,
		StateTransitions: metrics.StateTransitionCount,
		FinalState:       t.circuitBreaker.GetState(),
		TimeInState:      metrics.TimeInState,
		AvgResponseTimeMs: func() float64 {
			if successCount+failureCount > 0 {
				return float64(totalResponseTime) / float64(successCount+failureCount)
			}
			return 0.0
		}(),
	}, nil
}

// simulateRequest simulates a request with chaos injection
func (t *CircuitBreakerTester) simulateRequest(ctx context.Context) error {
	t.mu.Lock()
	config := *t.config
	t.mu.Unlock()

	// Inject latency if enabled
	if config.InjectLatency {
		latency := time.Duration(config.LatencyMs) * time.Millisecond

		select {
		case <-time.After(latency):
			// Latency injected
		case <-ctx.Done():
			return ctx.Err()
		}
	}

	// Inject connection errors if enabled
	if config.InjectConnectionErrors && t.rand.Float64() < config.ConnectionErrorRate {
		return errors.New("simulated connection error")
	}

	// Inject timeouts if enabled
	if config.InjectTimeouts && t.rand.Float64() < config.TimeoutRate {
		return errors.New("simulated timeout")
	}

	// Inject failures if enabled
	if config.InjectFailures && t.rand.Float64() < config.FailureRate {
		return errors.New("simulated failure")
	}

	return nil
}

// ResilienceTestSuite is a test suite for resilience patterns
type ResilienceTestSuite struct {
	// CircuitBreakerTester is the circuit breaker tester
	circuitBreakerTester *CircuitBreakerTester
	// Config is the chaos testing configuration
	config *ChaosTestingConfig
	// Mutex for protecting config updates
	mu sync.Mutex
	// Logger
	logger *zap.Logger
}

// NewResilienceTestSuite creates a new resilience test suite
func NewResilienceTestSuite() *ResilienceTestSuite {
	logger, _ := zap.NewProduction()

	return &ResilienceTestSuite{
		config: DefaultChaosTestingConfig(),
		logger: logger,
	}
}

// WithCircuitBreaker adds a circuit breaker tester to the test suite
func (s *ResilienceTestSuite) WithCircuitBreaker(circuitBreaker *CircuitBreaker) *ResilienceTestSuite {
	config := DefaultChaosTestingConfig()
	s.circuitBreakerTester = NewCircuitBreakerTester(circuitBreaker, *config)
	return s
}

// UpdateConfig updates the chaos testing configuration
func (s *ResilienceTestSuite) UpdateConfig(config ChaosTestingConfig) {
	s.mu.Lock()
	defer s.mu.Unlock()

	s.config = &config

	if s.circuitBreakerTester != nil {
		s.circuitBreakerTester.UpdateConfig(config)
	}
}

// RunCircuitBreakerTest runs a circuit breaker test
func (s *ResilienceTestSuite) RunCircuitBreakerTest(ctx context.Context, requestCount int64) (*CircuitBreakerTestResult, error) {
	if s.circuitBreakerTester == nil {
		return nil, errors.New("circuit breaker tester not configured")
	}

	return s.circuitBreakerTester.RunTest(ctx, requestCount)
}

// VerifyCircuitBreakerBehavior verifies circuit breaker behavior
func (s *ResilienceTestSuite) VerifyCircuitBreakerBehavior(ctx context.Context) (bool, error) {
	if s.circuitBreakerTester == nil {
		return false, errors.New("circuit breaker tester not configured")
	}

	// Configure chaos to inject failures
	s.circuitBreakerTester.UpdateConfig(ChaosTestingConfig{
		FailureRate:    1.0, // 100% failure rate
		InjectFailures: true,
	})

	// Run test with enough requests to trigger circuit breaker
	result, err := s.circuitBreakerTester.RunTest(ctx, 10)
	if err != nil {
		return false, err
	}

	// Verify circuit breaker opened
	if result.FinalState != CircuitBreakerStateOpen {
		return false, nil
	}

	// Configure chaos to stop injecting failures
	s.circuitBreakerTester.UpdateConfig(ChaosTestingConfig{
		FailureRate:    0.0,
		InjectFailures: false,
	})

	// Wait for circuit breaker to transition to half-open
	select {
	case <-time.After(31 * time.Second):
		// Time passed
	case <-ctx.Done():
		return false, ctx.Err()
	}

	// Run test with enough requests to close circuit breaker
	result, err = s.circuitBreakerTester.RunTest(ctx, 10)
	if err != nil {
		return false, err
	}

	// Verify circuit breaker closed
	return result.FinalState == CircuitBreakerStateClosed, nil
}

// ResilienceTestReport contains a report of a resilience test
type ResilienceTestReport struct {
	// TestName is the test name
	TestName string
	// DurationMs is the test duration in milliseconds
	DurationMs int64
	// Success indicates the test result (success/failure)
	Success bool
	// Details contains test details
	Details map[string]string
	// CircuitBreakerResults contains circuit breaker test results
	CircuitBreakerResults *CircuitBreakerTestResult
}

// ResilienceTestRunner runs resilience tests
type ResilienceTestRunner struct {
	// TestSuites contains test suites
	testSuites map[string]*ResilienceTestSuite
	// TestReports contains test reports
	testReports []*ResilienceTestReport
	// Mutex for protecting test reports
	mu sync.Mutex
	// Logger
	logger *zap.Logger
}

// NewResilienceTestRunner creates a new resilience test runner
func NewResilienceTestRunner() *ResilienceTestRunner {
	logger, _ := zap.NewProduction()

	return &ResilienceTestRunner{
		testSuites:  make(map[string]*ResilienceTestSuite),
		testReports: make([]*ResilienceTestReport, 0),
		logger:      logger,
	}
}

// AddTestSuite adds a test suite
func (r *ResilienceTestRunner) AddTestSuite(name string, suite *ResilienceTestSuite) {
	r.testSuites[name] = suite
}

// RunAllTests runs all tests
func (r *ResilienceTestRunner) RunAllTests(ctx context.Context) ([]*ResilienceTestReport, error) {
	r.mu.Lock()
	defer r.mu.Unlock()

	reports := make([]*ResilienceTestReport, 0, len(r.testSuites))

	for name, suite := range r.testSuites {
		startTime := time.Now()

		success, err := suite.VerifyCircuitBreakerBehavior(ctx)

		endTime := time.Now()

		durationMs := endTime.Sub(startTime).Milliseconds()

		details := make(map[string]string)
		if err != nil {
			details["error"] = err.Error()
		}

		report := &ResilienceTestReport{
			TestName:   name,
			DurationMs: durationMs,
			Success:    err == nil && success,
			Details:    details,
		}

		reports = append(reports, report)
		r.testReports = append(r.testReports, report)
	}

	return reports, nil
}

// GetTestReports returns all test reports
func (r *ResilienceTestRunner) GetTestReports() []*ResilienceTestReport {
	r.mu.Lock()
	defer r.mu.Unlock()

	// Return a copy of the reports
	reports := make([]*ResilienceTestReport, len(r.testReports))
	copy(reports, r.testReports)

	return reports
}

// ClearTestReports clears all test reports
func (r *ResilienceTestRunner) ClearTestReports() {
	r.mu.Lock()
	defer r.mu.Unlock()

	r.testReports = make([]*ResilienceTestReport, 0)
}

// CreateTestCircuitBreaker creates a circuit breaker for testing
func CreateTestCircuitBreaker(serviceName string) *CircuitBreaker {
	config := CircuitBreakerConfig{
		ConsecutiveErrorsThreshold: 5,
		OpenToHalfOpenTimeoutMs:    30000,
		HalfOpenSuccessThreshold:   3,
	}

	return NewCircuitBreaker(serviceName, config)
}

// CreateTestSuiteForService creates a resilience test suite for a service
func CreateTestSuiteForService(serviceName string) *ResilienceTestSuite {
	circuitBreaker := CreateTestCircuitBreaker(serviceName)

	return NewResilienceTestSuite().WithCircuitBreaker(circuitBreaker)
}
